File: projects/python-test/ehl_protocol.py
"""
Minimal EHL (legacy dispenser) framing helpers.

Frame format (as used by the VB6 app in this repo):
  STX (1) + LEN (1) + ADDR (1) + CMD (1) + DATA (0..n) + CHK (1) + ETX (1)

Checksum is XOR of all bytes from STX up to last DATA byte
(i.e., everything except CHK and ETX).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Iterable, List, Optional, Tuple


STX_CONTROLLER = 0x10
STX_DISPENSER = 0x20
ETX = 0x36


CMD_NAMES: Dict[int, str] = {
    0x1E: "OK",
    0x25: "ERROR_DATA",
    0x2F: "STOP",
    0x45: "VOLUME",
    0x4B: "STATE",
    0x4C: "ERROR_QUERY",
    0x5C: "PRICE",
    0x69: "BLOCK",
    0x6A: "LINETEST",
    0x70: "PROG_VOLUME",
    0x75: "PROG_AMOUNT",
    0x77: "UNBLOCK",
    0x81: "RESET",
    0xA9: "PROG_PRC",
    0xC3: "PRODUCT_SELECT",
    0xC5: "TANKBIT",
}


def cmd_name(cmd: int) -> str:
    return CMD_NAMES.get(cmd & 0xFF, f"UNKNOWN_0x{cmd & 0xFF:02X}")


def hexdump(b: bytes) -> str:
    return " ".join(f"{x:02X}" for x in b)


def xor_checksum(frame_wo_chk_etx: bytes) -> int:
    x = 0
    for bb in frame_wo_chk_etx:
        x ^= bb
    return x & 0xFF


def build_frame(
    addr: int,
    cmd: int,
    data: bytes = b"",
    *,
    stx: int = STX_CONTROLLER,
    etx: int = ETX,
) -> bytes:
    if not (1 <= addr <= 255):
        raise ValueError(f"addr must be 1..255, got {addr}")
    cmd_b = cmd & 0xFF
    length = 6 + len(data)
    if length > 255:
        raise ValueError(f"frame too long: {length}")

    header = bytes([stx & 0xFF, length & 0xFF, addr & 0xFF, cmd_b])
    body = header + data
    chk = xor_checksum(body)
    return body + bytes([chk, etx & 0xFF])


@dataclass(frozen=True)
class EhlFrame:
    stx: int
    length: int
    addr: int
    cmd: int
    data: bytes
    checksum: int
    etx: int

    @property
    def cmd_name(self) -> str:
        return cmd_name(self.cmd)

    def to_bytes(self) -> bytes:
        data = bytes(self.data)
        return build_frame(self.addr, self.cmd, data, stx=self.stx, etx=self.etx)


def parse_one_frame(buf: bytes) -> Optional[Tuple[EhlFrame, int]]:
    """
    Try to parse a single frame at buf[0]. Returns (frame, consumed_bytes) or None if incomplete/invalid.
    This expects buf[0] is already aligned to an STX.
    """
    if len(buf) < 6:
        return None

    stx = buf[0]
    if stx not in (STX_CONTROLLER, STX_DISPENSER):
        return None

    length = buf[1]
    if length < 6 or length > 255:
        return None
    if len(buf) < length:
        return None
    if buf[length - 1] != ETX:
        return None

    addr = buf[2]
    cmd = buf[3]
    data_len = length - 6
    data = buf[4 : 4 + data_len]
    checksum = buf[length - 2]

    chk_calc = xor_checksum(buf[: length - 2])
    if chk_calc != checksum:
        return None

    frame = EhlFrame(
        stx=stx,
        length=length,
        addr=addr,
        cmd=cmd,
        data=data,
        checksum=checksum,
        etx=ETX,
    )
    return frame, length


def extract_frames(stream_bytes: bytes) -> Tuple[List[EhlFrame], bytes]:
    """
    Extract as many valid frames as possible from a byte stream.
    Returns (frames, remainder_bytes).
    """
    frames: List[EhlFrame] = []
    i = 0
    n = len(stream_bytes)

    while i < n:
        # Find next STX
        while i < n and stream_bytes[i] not in (STX_CONTROLLER, STX_DISPENSER):
            i += 1
        if i >= n:
            return frames, b""

        parsed = parse_one_frame(stream_bytes[i:])
        if parsed is None:
            # Could be incomplete; keep remainder from i
            return frames, stream_bytes[i:]
        frame, consumed = parsed
        frames.append(frame)
        i += consumed

    return frames, b""


def describe_frame(f: EhlFrame) -> str:
    data_hex = hexdump(f.data) if f.data else "<none>"
    return (
        f"STX=0x{f.stx:02X} LEN={f.length} ADDR={f.addr} "
        f"CMD=0x{f.cmd:02X}({f.cmd_name}) DATA={data_hex} "
        f"CHK=0x{f.checksum:02X} ETX=0x{f.etx:02X}"
    )


def bits8(x: int) -> str:
    return f"{x & 0xFF:08b}"


def interpret_state_byte(state_b: int) -> Dict[str, object]:
    """
    Interpret STATE (0x4B) using VB6 bit checks:
      - Mid(state_string, 7, 1) => open for delivery  => bit1 (0x02)
      - Mid(state_string, 6, 1) => start button       => bit2 (0x04)
      - Mid(state_string, 5, 1) => automode           => bit3 (0x08)

    VB6 builds state_string using decimaltobinn(x(4)) (8-bit string, MSB first).
    """
    sb = state_b & 0xFF
    return {
        "raw": sb,
        "bits": bits8(sb),
        "open_for_delivery": bool(sb & 0x02),
        "startbutton_pressed": bool(sb & 0x04),
        "automode": bool(sb & 0x08),
        # Remaining bits are not labeled in VB6 in this repo
        "bit0": bool(sb & 0x01),
        "bit4": bool(sb & 0x10),
        "bit5": bool(sb & 0x20),
        "bit6": bool(sb & 0x40),
        "bit7": bool(sb & 0x80),
    }


def _vb_val_chr_byte(b: int) -> int:
    """
    VB6 uses Val(Chr(x(n))) for error levels. If the device sends ASCII digits,
    Val('3') -> 3. If it sends raw bytes, this is ambiguous; we fall back to numeric.
    """
    bb = b & 0xFF
    if 0x30 <= bb <= 0x39:
        return bb - 0x30
    return bb


_VB6_ERR_TEXT_NO: Dict[tuple[int, int], str] = {
    # Extracted from norgesgass_legacy/defs.bas -> logdisp_err()
    (1, 1): "Ingen kommunikasjon Display<-->CPU",
    (1, 2): "For mange kommunikasjonsfeil Display<-->CPU",
    (1, 3): "Intern feil Display",
    (2, 1): "Pulser ikke tilkoblet",
    (2, 2): "Feil rotasjon på pulser",
    (2, 3): "En pulserkanal mangler",
    (2, 4): "Feil serie på pulser",
    (2, 5): "Pulser buffer overflow",
    (2, 6): "LPG flow for høy",
    (3, 1): "Output overload(Para 10)",
    (3, 2): "Output control failure",
    (3, 3): "Startknapp aktivert under oppstart",
    (3, 4): "No load detected",
    (3, 5): "Termisk pumpebeskyttelse aktivert",
    (4, 1): "Minnefeil system",
    (4, 2): "Reset aktivert på hovedkort",
    (4, 3): "Strømbrudd",
    (4, 4): "Intern kommunikasjon CPU<-->Mainstream",
    (4, 5): "Calculations owerflow",
    (4, 7): "Brownout reset- for lite strøm til prosessor",
    (4, 8): "Ingen svar fra CPU",
    (5, 1): "Ingen Rs485 kommunikasjon",
    (6, 1): "Fylling har pågått for lenge (Para 22)",
    (6, 2): "For lang tid uten pulser ( Para 24)",
    (6, 3): "Flow for høy(Para 45)",
    (6, 4): "Maksimal grense for beløp nådd",
    (6, 6): "Pris er satt til 0.00",
    (6, 7): "Flow for liten (Para 48)",
    (6, 8): "Feil transaksjon state",
}


def interpret_error_query(data: bytes) -> Dict[str, object]:
    """
    Interpret ERROR_QUERY (0x4C) reply using VB6 'logdisp_err' mapping.
    VB6 treats main/sub levels as Val(Chr(x(4))), Val(Chr(x(5))).
    """
    if len(data) < 2:
        return {"ok": False, "reason": "missing bytes", "raw": hexdump(data)}
    main = _vb_val_chr_byte(data[0])
    sub = _vb_val_chr_byte(data[1])
    msg_no = _VB6_ERR_TEXT_NO.get((main, sub))
    # Best-effort English: translation of the VB6 Norwegian text (not new protocol info).
    msg_en = None
    if msg_no:
        msg_en = msg_no  # default to NO if we don't translate explicitly
        if msg_no == "Pris er satt til 0.00":
            msg_en = "Price is set to 0.00"
        elif msg_no == "Ingen Rs485 kommunikasjon":
            msg_en = "No RS-485 communication"
        elif msg_no == "Startknapp aktivert under oppstart":
            msg_en = "Start button active during startup"
        elif msg_no == "Strømbrudd":
            msg_en = "Power failure"
        elif msg_no == "Termisk pumpebeskyttelse aktivert":
            msg_en = "Thermal pump protection active"
    return {
        "ok": True,
        "main": main,
        "sub": sub,
        "message_no": msg_no,
        "message_en": msg_en,
        "raw": hexdump(data[:2]),
    }


def interpret_price_bytes(data: bytes) -> Optional[str]:
    # VB6: Chr(x(7)) & Chr(x(6)) & "." & Chr(x(5)) & Chr(x(4))
    if len(data) < 4:
        return None
    try:
        return f"{chr(data[3])}{chr(data[2])}.{chr(data[1])}{chr(data[0])}"
    except Exception:
        return None


def interpret_volume_bytes(data: bytes) -> Optional[str]:
    # VB6: Chr(x(8)) & Chr(x(7)) & Chr(x(6)) & \",\" & Chr(x(5)) & Chr(x(4))
    if len(data) < 5:
        return None
    try:
        return f"{chr(data[4])}{chr(data[3])}{chr(data[2])}.{chr(data[1])}{chr(data[0])}"
    except Exception:
        return None


def interpret_frame(f: EhlFrame) -> Optional[str]:
    """
    Return a best-effort human interpretation (based on VB6 code in this repo).
    Intended for field logs; never used for protocol validation.
    """
    if f.cmd == 0x4B and len(f.data) >= 1:
        s = interpret_state_byte(f.data[0])
        return (
            f"STATE raw=0x{s['raw']:02X} bits={s['bits']} "
            f"open_for_delivery={s['open_for_delivery']} "
            f"startbutton_pressed={s['startbutton_pressed']} "
            f"automode={s['automode']}"
        )
    if f.cmd == 0x4C:
        e = interpret_error_query(f.data)
        if not e.get("ok"):
            return f"ERROR_QUERY {e}"
        msg_no = e.get("message_no")
        msg_en = e.get("message_en")
        if msg_no and msg_en and msg_en != msg_no:
            return f"ERROR main={e['main']} sub={e['sub']} NO='{msg_no}' EN='{msg_en}'"
        if msg_no:
            return f"ERROR main={e['main']} sub={e['sub']} '{msg_no}'"
        return f"ERROR main={e['main']} sub={e['sub']} (unknown)"
    if f.cmd == 0x5C:
        p = interpret_price_bytes(f.data)
        if p:
            return f"PRICE {p}"
    if f.cmd == 0x45:
        v = interpret_volume_bytes(f.data)
        if v:
            return f"VOLUME {v}"
    return None

File: projects/pump-steering/EhlProtocol.cs
namespace pump_steering
{
    public class EhlFrame(byte[] data)
    {
        public EhlFrame() : this(data: [])
        {
        }

        public byte Stx { get; init; }
        public byte Length { get; init; }
        public byte Addr { get; init; }
        public byte Cmd { get; init; }
        public byte[] Data { get; set; } = data;
        public byte Checksum { get; init; }
        public byte Etx { get; init; }

        public override string ToString()
        {
            string dataHex = Data.Length > 0 ? BitConverter.ToString(Data).Replace("-", " ") : "<none>";
            return $"STX=0x{Stx:X2} LEN={Length} ADDR={Addr} CMD=0x{Cmd:X2} DATA={dataHex} CHK=0x{Checksum:X2} ETX=0x{Etx:X2}";
        }
    }

    public enum ParseResult
    {
        Success,
        Incomplete,
        Invalid
    }

    public static class EhlProtocol
    {
        private const byte STX_CONTROLLER = 0x10;
        public const byte STX_DISPENSER = 0x20;
        private const byte ETX = 0x36;

        public const byte CMD_STATE = 0x4B;
        public const byte CMD_VOLUME = 0x45;
        public const byte CMD_PRICE = 0x5C;
        public const byte CMD_ERROR_QUERY = 0x4C;
        public const byte CMD_PRODUCT_SELECT = 0xC3;
        public const byte CMD_PROG_PRC = 0xA9;
        public const byte CMD_UNBLOCK = 0x77;
        public const byte CMD_BLOCK = 0x69;
        public const byte CMD_RESET = 0x81;

        private static byte ComputeChecksum(byte[] buffer, int length)
        {
            byte x = 0;
            // XOR from start (STX) up to the byte before checkum
            for (int i = 0; i < length; i++)
            {
                x ^= buffer[i];
            }
            return x;
        }

        public static byte[] BuildFrame(byte addr, byte cmd, byte[]? data = null)
        {
            data ??= [];
            
            // Frame structure: STX (1) + LEN (1) + ADDR (1) + CMD (1) + DATA (N) + CHK (1) + ETX (1)
            
            var frameLen = 6 + data.Length;
            if (frameLen > 255) throw new ArgumentException("Frame too long");

            byte[] frame = new byte[frameLen];
            frame[0] = STX_CONTROLLER;
            frame[1] = (byte)frameLen;
            frame[2] = addr;
            frame[3] = cmd;
            
            if (data.Length > 0)
            {
                Array.Copy(data, 0, frame, 4, data.Length);
            }

            // Checksum over indices 0 to (frameLen - 3) inclusive (so length - 2 bytes)
            frame[frameLen - 2] = ComputeChecksum(frame, frameLen - 2);
            frame[frameLen - 1] = ETX;

            return frame;
        }

        public static (ParseResult Result, EhlFrame Frame, int Consumed) ParseOneFrame(byte[] buffer, int offset, int count)
        {
            if (count < 6) return (ParseResult.Incomplete, null, 0)!;

            byte stx = buffer[offset];
            if (stx != STX_CONTROLLER && stx != STX_DISPENSER) return (ParseResult.Invalid, null, 0)!; 

            byte length = buffer[offset + 1];
            if (length < 6) return (ParseResult.Invalid, null, 0)!; // Invalid length byte
            if (count < length) return (ParseResult.Incomplete, null, 0)!; // Need more bytes

            if (buffer[offset + length - 1] != ETX) return (ParseResult.Invalid, null, 0)!; 

            byte checksum = buffer[offset + length - 2];
            
            // Validate checksum (XOR of 0 ... length-3)
            byte calculated = 0;
            for (int i = 0; i < length - 2; i++)
            {
                calculated ^= buffer[offset + i];
            }

            if (calculated != checksum) return (ParseResult.Invalid, null, 0)!;

            var frame = new EhlFrame
            {
                Stx = stx,
                Length = length,
                Addr = buffer[offset + 2],
                Cmd = buffer[offset + 3],
                Checksum = checksum,
                Etx = ETX
            };

            int dataLen = length - 6;
            frame.Data = new byte[dataLen];
            Array.Copy(buffer, offset + 4, frame.Data, 0, dataLen);

            return (ParseResult.Success, frame, length);
        }

        // Data Interpretation Helpers

        public static string? InterpretVolume(byte[]? data)
        {
            if (data == null || data.Length < 5) return null;
            try 
            {
                var chars = new char[6];
                chars[0] = (char)data[4];
                chars[1] = (char)data[3];
                chars[2] = (char)data[2];
                chars[3] = '.';
                chars[4] = (char)data[1];
                chars[5] = (char)data[0];
                return new string(chars);
            }
            catch { return null; }
        }

        public static string? InterpretPrice(byte[]? data)
        {
            if (data == null || data.Length < 4) return null;
            try
            {
                var chars = new char[5];
                chars[0] = (char)data[3];
                chars[1] = (char)data[2];
                chars[2] = '.';
                chars[3] = (char)data[1];
                chars[4] = (char)data[0];
                return new string(chars);
            }
            catch { return null; }
        }

        public static byte[] EncodePrice(string priceStr)
        {
             var p = priceStr.Trim();
             // Validate format XX.XX
             if (p.Length != 5 || p[2] != '.') throw new ArgumentException("Price must be XX.XX");
             var digits = p.Replace(".", "");
             return digits.Length != 4 ? throw new ArgumentException("Price must have 4 digits") : [(byte)digits[3], (byte)digits[2], (byte)digits[1], (byte)digits[0]];
        }

        public static (bool OpenForDelivery, bool StartButtonPressed, bool AutoMode) InterpretState(byte stateByte)
        {
            // bit1 (0x02) -> open
            // bit2 (0x04) -> start
            // bit3 (0x08) -> auto
            
            var open = (stateByte & 0x02) != 0;
            var start = (stateByte & 0x04) != 0;
            var auto = (stateByte & 0x08) != 0;
            return (open, start, auto);
        }
    }
}
